### ISSUE TITLE
`cider-repl-set-ns' from cider broken with weasel

### ISSUE THREAD

**Issue by the-kenny** [tomjakubowski/weasel#26]
When using weasel with cider in recent clojurescript/cider-nrepl/cider setups, ```C-c M-n``` (```cider-repl-set-ns```) seems to broken. It switches to the namespace, but evaling anything in the repl will switch back to the ```user``` namespace.

This works fine when using cider without weasel, and I suppose it's some kind of mismatch between the expectations of cider and weasel. This used to work some weeks ago.

**Comment by sritchie** [comment]
I just upgraded cider and am running into this too. I'm on Cider 0.8.2 and

[org.clojure/clojurescript "0.0-2371"]

Here's what I get when I set debug messages on:

```clojure
(--->
  ns  "racehub.client.session.facebook"
  op  "eval"
  session  "02b1c99c-fc90-4b20-9c67-3de596fcd42f"
  code  "10\n"
  id  "241"
)
(<-
  id  "241"
  ns  "cljs.user"
  session  "02b1c99c-fc90-4b20-9c67-3de596fcd42f"
  value  "10"
)
(<-
  id  "241"
  session  "02b1c99c-fc90-4b20-9c67-3de596fcd42f"
  status  ("done")
)
```

**Comment by bbatsov** [comment]
The response is clearly wrong - the request has one `ns` in it and the response has another one. This has to be fixed in weasel.

**Comment by the-kenny** [comment]
Note: This is still an issue with the experimental code of @mfikes which supports the new REPL API of ```clojurescript-0.0-2665```.

**Comment by bbatsov** [comment]
Btw, thinking a bit more about it - this is likely an issue in piggieback as weasel actually uses it. @cemerick is this some piggieback bug?

**Comment by cemerick** [comment]
I'm using piggieback 0.1.5-SNAPSHOT with cider 0.6.0alpha, and cider-repl-set-ns works as expected.

**Comment by bbatsov** [comment]
@cemerick 0.6 is way too old. Back then `ns` switching actually invoked `in-ns`, but afterwards the ns tracking was done internally by cider (the ns gets updated when an eval request results in a new `ns` in the response).

**Comment by cemerick** [comment]
It's old, yes, but it works well, so I've never felt the need to upgrade. "tracking nd internally" seems like a bad idea, for the reasons I discussed on the cider issue being discussed concurrently...

**Comment by bbatsov** [comment]
I have no problem changing this in cider, but still I think that piggieback's behaviour should be consistent with the standard nREPL eval's behaviour. I wrote a bit more on the subject just now in the other ticket.

**Comment by sritchie** [comment]
Try it with Cider 0.8.2.

—

**Comment by cemerick** [comment]
@sritchie Change a working setup? Are you _mad_, dude? :-)

**Comment by sritchie** [comment]
@cemerick I was a complete madman running on a SNAPSHOT build of cider for a while... the SNAPSHOT jar in my `profile.clj` would auto update while emacs stayed put. I'd have maddening bugs for like two days until someone pushed a fix to master :)

**Comment by the-kenny** [comment]
Isn't it that ```interruptible-eval``` (the middleware) optionally takes a ```:ns``` map entry? If so, piggieback should support this.

**Comment by cichli** [comment]
@the-kenny piggieback does support this (see [here](https://github.com/cemerick/piggieback/blob/master/src/cemerick/piggieback.clj#L89)), it seems the issue is that nREPL is somehow setting `*ns*` when the `:ns` entry is present, and CIDER is relying upon this behaviour.

The piggieback way makes more sense to me - the current ns is only changed if `in-ns` is called explicitly, and if you want to do an `eval` in a different namespace to the current ns you don't have to switch back afterwards.

**Comment by tomjakubowski** [comment]
This appears to be fixed using nREPL 0.2.7 and the latest CIDER on MELPA.

Thanks to everyone for tracking this down!
